File: kre-go/context.go
package kre

import (
	"path"
	"time"

	"github.com/nats-io/nats.go"
	"google.golang.org/protobuf/proto"
	"google.golang.org/protobuf/types/known/anypb"

	"github.com/konstellation-io/kre-runners/kre-go/v4/config"
	"github.com/konstellation-io/kre-runners/kre-go/v4/mongodb"
	"github.com/konstellation-io/kre/libs/simplelogger"
)

var (
	saveMetricTimeout = 1 * time.Second
	saveDataTimeout   = 1 * time.Second
	getDataTimeout    = 1 * time.Second
)

const (
	defaultValue = ""
)

type PublishMsgFunc = func(response proto.Message, reqMsg *KreNatsMessage, msgType MessageType, channel string) error
type PublishAnyFunc = func(response *anypb.Any, reqMsg *KreNatsMessage, msgType MessageType, channel string)

type HandlerContextParams struct {
	Cfg                config.Config
	NC                 *nats.Conn
	MongoManager       mongodb.Manager
	Logger             *simplelogger.SimpleLogger
	PublishMsg         PublishMsgFunc
	PublishAny         PublishAnyFunc
	ContextObjectStore *contextObjectStore
}

type HandlerContext struct {
	cfg         config.Config
	publishMsg  PublishMsgFunc
	publishAny  PublishAnyFunc
	reqMsg      *KreNatsMessage
	Logger      *simplelogger.SimpleLogger
	ObjectStore *contextObjectStore
	Prediction  *contextPrediction
	Measurement *contextMeasurement
	DB          *contextDatabase
}

func NewHandlerContext(params *HandlerContextParams) *HandlerContext {
	return &HandlerContext{
		cfg:         params.Cfg,
		publishMsg:  params.PublishMsg,
		publishAny:  params.PublishAny,
		Logger:      params.Logger,
		ObjectStore: params.ContextObjectStore,
		Prediction:  NewContextPrediction(params.Cfg, params.NC, params.Logger),
		Measurement: NewContextMeasurement(params.Cfg, params.Logger),
		DB:          NewContextDatabase(params.Cfg, params.NC, params.MongoManager, params.Logger),
	}
}

// Path will return the relative path given as an argument as a full path.
func (c *HandlerContext) Path(relativePath string) string {
	return path.Join(c.cfg.BasePath, relativePath)
}

// GetRequestID will return the payload's original request ID.
func (c *HandlerContext) GetRequestID() string {
	return c.reqMsg.RequestId
}

// SendOutput will send a desired typed proto payload to the node's subject.
// By specifying a channel, the message will be sent to that subject's subtopic.
//
// SendOutput converts the proto message into an any type. This means the following node will recieve
// an any type protobuf.
//
// GRPC requests can only be answered once. So once the entrypoint has been replied by the exitpoint,
// all following replies to the entrypoint from the same request will be ignored.
func (c *HandlerContext) SendOutput(response proto.Message, channelOpt ...string) error {
	return c.publishMsg(response, c.reqMsg, MessageType_OK, c.getOptionalString(channelOpt))
}

// SendAny will send any type of proto payload to the node's subject.
// By specifying a channel, the message will be sent to that subject's subtopic.
//
// As a difference from SendOutput, SendAny will not convert your proto structure.
//
// Use this function when you wish to simply redirect your node's payload without unpackaging.
// Once the entrypoint has been replied, all following replies to the entrypoint will be ignored.
func (c *HandlerContext) SendAny(response *anypb.Any, channelOpt ...string) {
	c.publishAny(response, c.reqMsg, MessageType_OK, c.getOptionalString(channelOpt))
}

// SendEarlyReply works as the SendOutput functionality
// with the addition of typing this message as an early reply.
func (c *HandlerContext) SendEarlyReply(response proto.Message, channelOpt ...string) error {
	return c.publishMsg(response, c.reqMsg, MessageType_EARLY_REPLY, c.getOptionalString(channelOpt))
}

// SendEarlyExit works as the SendOutput functionality
// with the addition of typing this message as an early exit.
func (c *HandlerContext) SendEarlyExit(response proto.Message, channelOpt ...string) error {
	return c.publishMsg(response, c.reqMsg, MessageType_EARLY_EXIT, c.getOptionalString(channelOpt))
}

func (c *HandlerContext) getOptionalString(values []string) string {
	if len(values) > 0 {
		return values[0]
	}
	return defaultValue
}

// IsMessageOK returns true if the incoming message is of message type OK.
func (c *HandlerContext) IsMessageOK() bool {
	return c.reqMsg.MessageType == MessageType_OK
}

// IsMessageError returns true if the incoming message is of message type ERROR.
func (c *HandlerContext) IsMessageError() bool {
	return c.reqMsg.MessageType == MessageType_ERROR
}

// IsMessageEarlyReply returns true if the incoming message is of message type EARLY REPLY.
func (c *HandlerContext) IsMessageEarlyReply() bool {
	return c.reqMsg.MessageType == MessageType_EARLY_REPLY
}

// IsMessageEarlyExit returns true if the incoming message is of message type EARLY EXIT.
func (c *HandlerContext) IsMessageEarlyExit() bool {
	return c.reqMsg.MessageType == MessageType_EARLY_EXIT
}

File: kre-py/src/context.py
import os
from logging import Logger
from typing import Any, Awaitable, Callable

from google.protobuf.message import Message as ProtobufMessage
from nats.aio.client import Client as NatsClient
from nats.js.client import JetStreamContext
from pymongo import MongoClient

from config import Config
from context_configuration import ContextConfiguration
from context_data import ContextData
from context_measurement import ContextMeasurement
from context_objectstore import ContextObjectStore
from context_prediction import ContextPrediction
from kre_nats_msg_pb2 import EARLY_EXIT, EARLY_REPLY, ERROR, OK, KreNatsMessage, MessageType

PublishMsgFunc = Callable[[ProtobufMessage, Any, MessageType, str], Awaitable]
PublishAnyFunc = Callable[[ProtobufMessage, Any, MessageType, str], Awaitable]

DEFAULT_CHANNEL = ""


class HandlerContext:
    def __init__(
        self,
        config: Config,
        nc: NatsClient,
        js: JetStreamContext,
        mongo_conn: MongoClient,
        logger: Logger,
        publish_msg: PublishMsgFunc,
        publish_any: PublishAnyFunc,
        configuration: ContextConfiguration,
    ):
        self.__config__: Config = config
        self.__publish_msg__ = publish_msg
        self.__publish_any__ = publish_any
        self.__request_msg__: KreNatsMessage
        self.logger = logger
        self.prediction = ContextPrediction(config, nc, logger)
        self.measurement = ContextMeasurement(config, logger)
        self.db = ContextData(config, nc, js, mongo_conn, logger)
        self.configuration = configuration

        if config.nats_object_store is not None:
            self.object_store = ContextObjectStore(config, logger)

    def path(self, relative_path):
        return os.path.join(self.__config__.base_path, relative_path)

    def set_request_msg(self, request_msg):
        self.__request_msg__ = request_msg

    def get_request_id(self) -> str:
        """get_request_id will return the payload's original request ID."""
        return str(self.__request_msg__.request_id)

    async def send_output(self, message: ProtobufMessage, channel: str = DEFAULT_CHANNEL) -> None:
        """
        send_output will send a desired typed proto payload to the node's subject.
        By specifying a channel, the message will be sent to that subject's subtopic.

        send_output converts the proto message into an any type. This means the following node will recieve
        an any type protobuf.

        GRPC requests can only be answered once. So once the entrypoint has been replied by the exitpoint,
        all following replies to the entrypoint from the same request will be ignored.

        :param  message:  The message to publish.
        :param channel (optional):  The subsubject channel where the message will be published.
        """
        await self.__publish_msg__(message, self.__request_msg__, OK, channel)

    async def send_any(self, message: ProtobufMessage, channel: str = DEFAULT_CHANNEL) -> None:
        """
        send_any will send any type of proto payload to the node's subject.
        By specifying a channel, the message will be sent to that subject's subtopic.

        As a difference from SendOutput, send_any will not convert your proto structure.

        Use this function when you wish to simply redirect your node's payload without unpackaging.
        Once the entrypoint has been replied, all following replies to the entrypoint will be ignored.

        :param  message:  The message to publish.
        :param channel (optional):  The subsubject channel where the message will be published.
        """
        await self.__publish_any__(message, self.__request_msg__, OK, channel)

    async def send_early_reply(
        self, message: ProtobufMessage, channel: str = DEFAULT_CHANNEL
    ) -> None:
        """
        send_early_reply works as the SendOutput functionality
        with the addition of typing this message as an early reply.

        :param  message:  The message to publish.
        :param channel (optional):  The subsubject channel where the message will be published.
        """
        await self.__publish_msg__(message, self.__request_msg__, EARLY_REPLY, channel)

    async def send_early_exit(
        self, message: ProtobufMessage, channel: str = DEFAULT_CHANNEL
    ) -> None:
        """
        send_early_exit works as the SendOutput functionality
        with the addition of typing this message as an early exit.

        :param  message:  The message to publish.
        :param channel (optional):  The subsubject channel where the message will be published.
        """
        await self.__publish_msg__(message, self.__request_msg__, EARLY_EXIT, channel)

    def is_message_ok(self) -> bool:
        """is_message_ok returns true if incoming message is of message type OK."""
        return self.__request_msg__.message_type == OK

    def is_message_error(self) -> bool:
        """is_message_error returns true if incoming message is of message type ERROR."""
        return self.__request_msg__.message_type == ERROR

    def is_message_early_reply(self) -> bool:
        """is_message_early_reply returns true if incoming message is of message type EARLY_REPLY."""
        return self.__request_msg__.message_type == EARLY_REPLY

    def is_message_early_exit(self) -> bool:
        """is_message_early_exit returns true if incoming message is of message type EARLY_EXIT."""
        return self.__request_msg__.message_type == EARLY_EXIT
